crates/interpreter/src/valid.rs

```diff
@@ -503,22 +507,53 @@ impl<'a, 'm> Expr<'a, 'm> {
         match instr {
             Unreachable => self.stack_polymorphic(),
             Nop => (),
-            Block(b) => self.push_label(self.blocktype(&b)?, LabelKind::Block)?,
-            Loop(b) => self.push_label(self.blocktype(&b)?, LabelKind::Loop)?,
+            Block(b) => self.push_label(self.blocktype(&b)?, LabelKind::Block, None, false)?,
+            Loop(b) => self.push_label(self.blocktype(&b)?, LabelKind::Loop, None, false)?,
             If(b) => {
                 self.pop_check(ValType::I32)?;
-                self.push_label(self.blocktype(&b)?, LabelKind::If)?;
+                self.push_label(
+                    self.blocktype(&b)?,
+                    LabelKind::If,
+                    Some(self.side_table.len()),
+                    false,
+                )?;
+                self.side_table.push(SideTableEntryView {
+                    delta_ip: -self.instr_idx,
+                    // TODO(dev/fast-interp): implement below.
+                    delta_stp: 0,
+                    val_cnt: 0,
+                    pop_cnt: 0,
+                });
             }
             Else => {
+                let side_table_idx = self.immutable_label().side_table_idx.unwrap();
+                let cur_instr_idx = self.instr_idx;
+                self.side_table[side_table_idx].delta_ip += self.instr_idx + 1;
+
                 let label = self.label();
+                label.is_block_under_else = true;
                 check(core::mem::replace(&mut label.kind, LabelKind::Block) == LabelKind::If)?;
                 let FuncType { params, results } = label.type_;
                 self.pops(results)?;
                 check(self.stack().is_empty())?;
                 self.label().polymorphic = false;
                 self.pushs(params);
+
+                self.side_table.push(SideTableEntryView {
+                    delta_ip: -cur_instr_idx,
+                    // TODO(dev/fast-interp): implement below.
+                    delta_stp: 0,
+                    val_cnt: 0,
+                    pop_cnt: 0,
+                });
+            }
+            End => {
+                if self.immutable_label().is_block_under_else {
+                    let side_table_idx = self.immutable_label().side_table_idx.unwrap() + 1;
+                    self.side_table[side_table_idx].delta_ip += self.instr_idx;
+                }
```

Comment:
You can see that this code is unreachable from the macro below. The `End` instruction is handled by `end_label()`.